### PR TITLE
feat: reduce memory usage of `Type` by avoiding re-allocations

### PR DESCRIPTION
**Pull request by magnus-madsen**
Please double check my work :)

Saves 20MB!

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala [hunk 1]
```diff
@@ -62,13 +62,39 @@ object Scheme {
...
+        // Performance: Reuse tpe0.
+        tpe0
+
+      case Type.Apply(tpe1, tpe2, _) =>
+        val t1 = visitType(tpe1)
+        val t2 = visitType(tpe2)
+        // Performance: Reuse tpe0 or tpe1/tpe2 individually.
+        if ((t1 eq tpe1) && (t2 eq tpe2)) {
+          tpe0
+        } else if (t1 eq tpe1) {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:83]: These cases are redundant, right?

File: main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala [hunk 2]
```diff
@@ -154,8 +160,19 @@ object Monomorpher {
           case (Type.Apply(Type.Cst(TypeConstructor.SchemaRowExtend(label), _), tpe, _), rest) => mkSchemaExtendSorted(label, tpe, rest, loc)
 
           // Else just apply.
-          case (x, y) => Type.Apply(x, y, loc)
+          case (x, y) =>
+            // Performance: Reuse tpe0 or x/y individually.
+            if ((x eq t1) && (y eq t2)) {
+              tpe0
+            } else if (x eq t1) {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala:167]: Ditto

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:83]: What do you mean by redundant?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:83]: Semantically they are redundant compared to the last case, but this is where memory is saved.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:83]: We reuse one subtree then, right?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:83]: Oh, I see what you mean. I should simplify then.

**Review by JonathanStarup — COMMENTED** [review]
Nothing besides the notes matt had

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala]: Fixed.

**Review by mlutze — APPROVED** [review]

**Pull request merged**